spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkExpressions.java

```
@@ -0,0 +1,322 @@
+        case EQ:
+        case NOT_EQ:
+          BinaryExpression binary = (BinaryExpression) expr;
+          return convert(op, binary.left(), binary.right());
+        case NOT:
+          org.apache.iceberg.expressions.Expression child = convert(((Not) expr).child());
+          if (child != null) {
+            return not(child);
+          }
+          return null;
+        case AND:
+          And andExpr = (And) expr;
+          org.apache.iceberg.expressions.Expression andLeft = convert(andExpr.left());
+          org.apache.iceberg.expressions.Expression andRight = convert(andExpr.right());
+          if (andLeft != null && andRight != null) {
+            return and(convert(andExpr.left()), convert(andExpr.right()));
+          }
+          return null;
+        case OR:
+          Or orExpr = (Or) expr;
+          org.apache.iceberg.expressions.Expression orLeft = convert(orExpr.left());
+          org.apache.iceberg.expressions.Expression orRight = convert(orExpr.right());
+          if (orLeft != null && orRight != null) {
+            return or(orLeft, orRight);
+          }
+          return null;
+        case IN:
+          if (expr instanceof In) {
+            In inExpr = (In) expr;
+            List<Object> literals = convertLiterals(JavaConverters.seqAsJavaListConverter(inExpr.list()).asJava());
+            if (literals != null) {
+              return convertIn(inExpr.value(), literals);
+            } else {
+              // if the list contained a non-literal, it can't be converted
+              return null;
+            }
+          } else if (expr instanceof InSet) {
+            InSet inExpr = (InSet) expr;
+            // expressions are already converted to Java objects
+            Set<Object> literals = JavaConverters.setAsJavaSetConverter(inExpr.hset()).asJava();
+            return convertIn(inExpr.child(), literals);
+          }
+          return null;
+        default:
+      }
+    }
+    // can't convert
+    return null;
```

what happens if we cant convert spark -> iceberg? Is it an error, something the client deals with or...

Naievly this seems like a case we would want to throw an exception

---

ok. I can throw unsupported exception with reason. So that the users can know.